server/src/main/java/org/elasticsearch/indices/recovery/MultiFileWriter.java

```diff
@@ -70,6 +74,41 @@ public void writeFileChunk(StoreFileMetadata fileMetadata, long position, Releas
         }
     }
 
+    public void writeFile(StoreFileMetadata fileMetadata, InputStream stream) throws Exception {
+        ensureOpen.run();
+        assert Transports.assertNotTransportThread("multi_file_writer");
+
+        String fileName = fileMetadata.name();
+        String tempFileName = getTempNameForFile(fileName);
+        if (tempFileNames.containsKey(tempFileName)) {
+            throw new IllegalStateException("output for file [" + fileName + "] has already been created");
+        }
+        tempFileNames.put(tempFileName, fileName);
+
+        incRef();
+        try(IndexOutput indexOutput = store.createVerifyingOutput(tempFileName, fileMetadata, IOContext.DEFAULT)) {
+            int bufferSize = Math.toIntExact(Math.min(DEFAULT_WRITE_FILE_BUFFER_SIZE, fileMetadata.length()));
+            byte[] buffer = new byte[bufferSize];
```

Comment:
I'm afraid we cannot recycle these buffers, since they would be a multi-page buffer and I think we cannot adapt them to the `InputStream` interface.

Reply:
Ok, I was wondering about something like this:

https://github.com/elastic/elasticsearch/blob/7714b7b6037112738a6b26f4b9326cdb4864330e/server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java#L1065

However on reflection I think we don't really need this, there's not normally that many files to restore. We don't recycle the buffers that get used in a regular restore. In investigating that I found that actually a better setting would be the repository setting `BlobStoreRepository#BUFFER_SIZE_SETTING` rather than the one I [suggested yesterday](https://github.com/elastic/elasticsearch/pull/76237#discussion_r685236629).

Reply:
Ah sorry, I misunderstood you. The original 128kb came from that setting 😅 . I'll change it 👍